### PR TITLE
IGNITE-16964  SQL API: Implement async SQL API

### PR DESCRIPTION
**Pull request by tledkov-gridgain**

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+        return batchPage.hasMore();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<Void> closeAsync() {
+        return cur.closeAsync().thenAccept((v) -> closeRun.run());
+    }
+
+    private class Page implements Iterable<SqlRow> {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:160]: Can the Page be static class?
Does it need an implicit reference to the parent object?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int columnIndex(@NotNull String columnName) {
+            return cur.metadata().fields().stream()
+                    .filter(fld -> fld.name().equals(columnName))
+                    .findFirst()
+                    .get()
+                    .order();
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:198]: Let's avoid streams here?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+        @Override
+        public Iterator<SqlRow> iterator() {
+            return batchPage.items().stream()
+                    .map(RowTuple::new)
+                    .map(SqlRow.class::cast)
+                    .iterator();
+        }
+    }
+
+    private class RowTuple implements SqlRow {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:172]:
```suggestion
    private class SqlRowImpl implements SqlRow {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    /** {@inheritDoc} */
+    @Override
+    public Publisher<Integer> executeBatchReactive(@Nullable Transaction transaction, Statement statement, BatchedArguments batch) {
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /**
+     * Session builder implementation.
+     */
+    public static class SessionBuilderImpl implements SessionBuilder {
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:219]: Let's make it a top-level class.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java (hunk 3 above)
Comment (tledkov-gridgain) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:172]: Fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java (hunk 2 above)
Comment (tledkov-gridgain) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:198]: Fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java (hunk 4 above)
Comment (tledkov-gridgain) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:219]: fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+    /** {@inheritDoc} */
+    @Override
+    public boolean hasMorePages() {
+        return batchPage.hasMore();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<Void> closeAsync() {
+        return cur.closeAsync().thenAccept((v) -> closeRun.run());
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:156]:
```suggestion
        return cur.closeAsync().thenRun(closeRun);
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Iterable<SqlRow> currentPage() {
+        if (!hasRowSet()) {
+            throw new NoRowSetExpectedException("Query hasn't result set: [type=" + cur.queryType() + ']');
+        }
+
+        return page;
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:125]:
```suggestion
        return () -> new TransformingIterator<>(batchPage.items().iterator(), SqlRowImpl::new);
```

and after that we don't need neither Page nor IteratorImpl classes

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+        /** {@inheritDoc} */
+        @Override
+        public String columnName(int columnIndex) {
+            return cur.metadata().fields().get(columnIndex).name();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public int columnIndex(@NotNull String columnName) {
+            return cur.metadata().fields().stream()
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:208]: does it make sense to prepare metadata once for the whole result and then just pass it to the row?

BTW, what if the column with the given name is not exists in the row?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+                    .findFirst()
+                    .get()
+                    .order();
+        }
+
+        /** {@inheritDoc} */
+        @SuppressWarnings("unchecked")
+        @Override
+        public <T> T valueOrDefault(@NotNull String columnName, T defaultValue) {
+            T ret = (T) row.get(columnIndex(columnName));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:219]: with that in mind it totally make sense to prepare meta...

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+            List<CompletableFuture<AsyncSqlCursor<List<Object>>>> futs = qryProc.queryAsync(ctx, schema, query, arguments);
+
+            futsToClose.addAll(futs);
+
+            if (futs.size() != 1) {
+                return CompletableFuture.failedFuture(new IgniteSqlException("Multiple statements aren't allowed."));
+            }
+
+            return futs.get(0).thenCompose(cur -> {
+                        futsToClose.remove(futs.get(0));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:182]: wrong indentation

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+        QueryContext ctx = QueryContext.of(transaction, new QueryTimeout(timeout, TimeUnit.NANOSECONDS));
+
+        try {
+            List<CompletableFuture<AsyncSqlCursor<List<Object>>>> futs = qryProc.queryAsync(ctx, schema, query, arguments);
+
+            futsToClose.addAll(futs);
+
+            if (futs.size() != 1) {
+                return CompletableFuture.failedFuture(new IgniteSqlException("Multiple statements aren't allowed."));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:178]: at this point you've just lost all opened cursors...

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    public CompletableFuture<Void> closeAsync() {
+        return CompletableFuture
+                .runAsync(busyLock::block)
+                .thenCompose(
+                        v0 -> {
+                            futsToClose.forEach(f -> f.cancel(true));
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:237]: cancellation of the future doesn't cancel the whole computation chain started. So we'll get a bunch of opened cursors here

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        /** {@inheritDoc} */
+        @Override
+        public long longValue(@NotNull String columnName) {
+            return (long) row.get(columnIndexChecked(columnName));
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public long longValue(int columnIndex) {
+            return 0;
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java:287]: did you miss this one?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+
+        assertEquals(2, r.columnCount());
+        assertEquals(0, r.columnIndex("COL_A"));
+        assertEquals(1, r.columnIndex("COL_B"));
+        assertEquals(-1, r.columnIndex("notExistColumn"));
+
+        assertEquals(1, r.intValue("COL_A"));
+        assertEquals(2, r.intValue("COL_B"));
+
+        assertThrowsWithCause(
```
Comment (korlov42) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/api/ItSqlAsynchronousApiTest.java:207]: let's add one more assertion to validate IndexOutOfBound is thrown when trying to access column by index which is not exists

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 14]
```diff
@@ -107,6 +107,27 @@ public static <T> T first(Collection<? extends T> col) {
         return col.iterator().next();
     }
 
+    /**
+     * Gets first element from given iterable or returns {@code null} if list is empty.
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:111]:
```suggestion
     * Returns first element from the given iterable or returns {@code null} if the list is empty.
```

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 15]
```diff
@@ -107,6 +107,27 @@ public static <T> T first(Collection<? extends T> col) {
...
+     *
+     * @param iterable Iterable to retrieve the first element.
+     * @param <T>  Type of the elements of the list.
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:114]:
```suggestion
     * @param <T> Type of the elements of the list.
```

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 16]
```diff
@@ -107,6 +107,27 @@ public static <T> T first(Collection<? extends T> col) {
...
+     * @return The first element of the given iterable or {@code null} in case the iterable is null or empty.
+     */
+    public static <T> T first(Iterable<? extends T> iterable) {
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:117]:
```suggestion
    public static <T> @Nullable T first(Iterable<? extends T> iterable) {
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java [hunk 17]
```diff
@@ -0,0 +1,424 @@
...
+    /** {@inheritDoc} */
+    @Override
+    public boolean hasRowSet() {
+        return cur.queryType() == SqlQueryType.QUERY || cur.queryType() == SqlQueryType.EXPLAIN;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long affectedRows() {
+        if (hasRowSet() || cur.queryType() == SqlQueryType.DDL) {
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/AsyncResultSetImpl.java]: I don't think so. An affectedRows should return amount of rows which were actually affected by the query, but explain returns a plan only without actual execution of this plan

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java [hunk 18]
```diff
@@ -228,6 +230,77 @@ public List<CompletableFuture<AsyncSqlCursor<List<Object>>>> queryAsync(QueryCon
...
+                    return new AsyncSqlCursorImpl<>(
+                            SqlQueryType.mapPlanTypeToSqlType(plan.type()),
+                            plan.metadata(),
+                            executionSrvc.executePlan(plan, ctx)
+                    );
+                });
+
+        stage.whenComplete((cur, ex) -> {
+            if (ex instanceof CancellationException) {
+                ctx.cancel().cancel();
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:297]: QueryCancel currently is not used

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProcessor.java [hunk 19]
```diff
@@ -48,6 +48,20 @@ public interface QueryProcessor extends IgniteComponent {
      * @return List of sql cursors.
      *
      * @throws IgniteException in case of an error.
-     * */
+     */
     List<CompletableFuture<AsyncSqlCursor<List<Object>>>> queryAsync(QueryContext context, String schemaName, String qry, Object... params);
+
+    /**
+     * Execute the single statement query with given schema name and parameters.
+     * If the query string contains more than one statement the IgniteException will be thrown.
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProcessor.java:56]:
```suggestion
     * 
     * <p>If the query string contains more than one statement the IgniteException will be thrown.
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProcessor.java [hunk 20]
```diff
@@ -48,6 +48,20 @@ public interface QueryProcessor extends IgniteComponent {
...
+     *
+     * @param context User query context.
+     * @param schemaName Schema name.
+     * @param qry Single statement SQL query .
+     * @param params Query parameters.
+     * @return List of sql cursors.
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryProcessor.java:62]: querySingle should return a single cursor

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java (hunk 18 above)
Comment (tledkov-gridgain) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java:297]: Is it a bug?

**Review by korlov42 — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java [hunk 21]
```diff
@@ -0,0 +1,283 @@
...
+
+                        if (!busyLock.enterBusy()) {
+                            return cur.closeAsync()
+                                    .thenCompose((v) -> CompletableFuture.failedFuture(new IgniteSqlException("Session is closed")));
+                        }
+
+                        try {
+                            cursToClose.add(cur);
+
+                            return cur.requestNextAsync(pageSize)
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:184]: Seems, cursor will not be removed from 'cursToClose' if 'requestNextAsync' call or the future fails.

Comment (tledkov-gridgain) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java:183]: Fixed, please take a look

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/IgniteTestUtils.java [hunk 22]
```diff
@@ -164,6 +164,36 @@ public static Object getFieldValue(Object target, Class<?> declaredClass, String
...
+     */
+    public static Object getFieldValue(Object obj, String fieldName) {
+        Class<?> cls = obj.getClass();
+        Exception ex = null;
+
+        while (cls != Object.class) {
+            try {
+                return getFieldValue(obj, obj.getClass(), fieldName);
+            } catch (Exception ex0) {
+                cls = cls.getSuperclass();
```
Comment (AMashenkov) [inline modules/core/src/test/java/org/apache/ignite/internal/testframework/IgniteTestUtils.java:182]: 'cls' is not used.
Do you actually mean this?
```suggestion
                return getFieldValue(obj, cls, fieldName);
            } catch (Exception ex0) {
                cls = cls.getSuperclass();
```

**Review by AMashenkov — APPROVED** [review]

File: modules/core/src/test/java/org/apache/ignite/internal/testframework/IgniteTestUtils.java (hunk 22 above)
Comment (tledkov-gridgain) [reply modules/core/src/test/java/org/apache/ignite/internal/testframework/IgniteTestUtils.java]: Fixed. Move test utils method from 2.x

**Pull request merged**